### PR TITLE
allow different tcp streams

### PR DESCRIPTION
**Pull request by spebern**
As suggested in https://github.com/spebern/async-ssh2/issues/4#issuecomment-578365811 this allows `ssh2-rs` to be used with other tcp streams, as long as a raw handle to the socket or file handle is obtainable.

### CODE REVIEW

File: src/session.rs [hunk 1]
```diff
@@ -263,31 +270,48 @@ impl Session {
...
+        let _ = inner.tcp.replace(Box::new(stream));
     }
 
-    /// Returns a reference to the stream that was associated with the Session
-    /// by the Session::handshake method.
-    pub fn tcp_stream(&self) -> Option<MappedMutexGuard<TcpStream>> {
+    /// Returns the raw fd of the stream that was associated with the Session by
+    /// the Session::handshake method.
+    #[cfg(unix)]
+    pub fn tcp_stream(&self) -> Option<RawFd> {
```
Comment (wez) [inline src/session.rs]: I know that I suggested this, but I wonder what you think about implementing `AsRawFd` and `AsRawSocket` on `Session` in the place of these methods?  Those traits don't allow for the stream to be `None`, but it's probably fine to have those implementations panic if used without assigning a stream.   I don't think anyone should really be using this method or those traits anyway.

File: src/session.rs [hunk 2]
```diff
@@ -263,31 +270,48 @@ impl Session {
...
-    /// You may use the tcp_stream() method to obtain a reference
-    /// to it later.
+    /// The session takes ownership of the stream provided.
+    /// You may use the tcp_stream() method to obtain the raw fd later.
+    ///
+    /// It is also highly recommended that the stream provided is not used
+    /// concurrently elsewhere for the duration of this session as it may
+    /// interfere with the protocol.
+    #[cfg(unix)]
+    pub fn set_tcp_stream<S: 'static + AsRawFd>(&mut self, stream: S) {
```
Comment (wez) [inline src/session.rs]: not a real concern, just an observation: I'm slightly surprised that the lifetime is needed here

File: src/session.rs (hunk 1 above)
Comment (spebern) [reply src/session.rs]: I think I like the panic approach with implementing those traits more. I'll change that.

File: src/session.rs (hunk 2 above)
Comment (spebern) [reply src/session.rs]: I think this is related to: https://stackoverflow.com/questions/29740488/parameter-type-may-not-live-long-enough .

So the bound doesn't allow that the trait object stored inside the session outlives everything it references to. I don't know if there is a better way, to me `'static` seems ok.

**Pull request merged**